torch/csrc/distributed/c10d/OpsImpl.cpp

```diff
@@ -49,6 +84,24 @@ TORCH_LIBRARY_IMPL(c10d, CPU, m) {
 TORCH_LIBRARY_IMPL(c10d, CUDA, m) {
   m.impl("broadcast_", broadcast_cuda_);
 }
+
+TORCH_LIBRARY_IMPL(c10d, CPU, m) {
+  m.impl("allreduce_", allreduce_cpu_);
+}
+
+// TODO: The SparseCPU/SparseCUDA dispatched methods are only used to support
+// sparse all_reduce in the Gloo backend
+TORCH_LIBRARY_IMPL(c10d, SparseCPU, m) {
+  m.impl("allreduce_", allreduce_cpu_);
+}
+
+TORCH_LIBRARY_IMPL(c10d, SparseCUDA, m) {
```

Comment:
Why do we have allreduce_cuda_ and allreduce_cpu_ when they look identical?

Reply:
Right now is setting up the skeleton to have dispatching route to different implementations for CPU/CUDA. Yes the implementations are identical for now, they will be updated in the future to perform `process_group->backend[<cpu>/<cuda>]->all_reduce`, but right now it is a no-op.